storage/src/mmr/mem.rs

```diff
@@ -409,124 +427,215 @@ impl<H: CHasher> Mmr<H> {
+    pub fn add_batched(&mut self, hasher: &mut impl Hasher<H>, element: &[u8]) -> Position {
+        let leaf_pos = self.size();
+        let digest = hasher.leaf_digest(leaf_pos, element);
+
+        // Compute the new parent nodes if any, and insert them into the MMR
+        // with a dummy digest, and add each to the dirty nodes set.
+        let nodes_needing_parents = nodes_needing_parents(self.peak_iterator())
+            .into_iter()
+            .rev();
+        self.nodes.push_back(digest);
+
+        let mut height = 1;
+        for _ in nodes_needing_parents {
+            let new_node_pos = self.size();
+            self.nodes.push_back(H::empty());
+            self.state.dirty_nodes.insert((new_node_pos, height));
+            height += 1;
+        }
+
+        leaf_pos
+    }
+
+    /// Batch update the digests of multiple retained leaves.
     ///
-    /// # Panics
+    /// # Errors
     ///
-    /// Panics if `self.pool` is None.
-    #[cfg(feature = "std")]
-    fn update_leaf_parallel<T: AsRef<[u8]> + Sync>(
+    /// Returns [Error::ElementPruned] if any of the leaves has been pruned.
+    pub fn update_leaf_batched<T: AsRef<[u8]> + Sync>(
         &mut self,
         hasher: &mut impl Hasher<H>,
         updates: &[(Position, T)],
-    ) {
-        let pool = self
-            .thread_pool
-            .as_ref()
-            .expect("pool must be non-None")
-            .clone();
-        pool.install(|| {
-            let digests: Vec<(Position, H::Digest)> = updates
-                .par_iter()
-                .map_init(
-                    || hasher.fork(),
-                    |hasher, (pos, elem)| {
-                        let digest = hasher.leaf_digest(*pos, elem.as_ref());
-                        (*pos, digest)
-                    },
-                )
-                .collect();
-
-            for (pos, digest) in digests {
-                let index = self.pos_to_index(pos);
-                self.nodes[index] = digest;
-                self.mark_dirty(pos);
+    ) -> Result<(), Error> {
+        for (pos, _) in updates {
+            if *pos < self.pruned_to_pos {
+                return Err(Error::ElementPruned(*pos));
             }
-        });
-    }
+        }
+
+        #[cfg(feature = "std")]
+        if updates.len() >= MIN_TO_PARALLELIZE && self.thread_pool.is_some() {
+            self.update_leaf_parallel(hasher, updates);
+        }
+
+        for (pos, element) in updates {
+            // Update the digest of the leaf node and mark its ancestors as dirty.
+            let digest = hasher.leaf_digest(*pos, element.as_ref());
+            let index = self.pos_to_index(*pos);
+            self.nodes[index] = digest;
+            self.mark_dirty(*pos);
+        }
 
-    /// Returns whether there are pending updates.
-    pub fn is_dirty(&self) -> bool {
-        !self.dirty_nodes.is_empty()
+        Ok(())
     }
 
     /// Merkleize any pending batched updates to have them reflected in the root.
```

Comment:
Maybe update this doc to explain the result is the clean MMR state?

Reply:
Changed to ` /// Convert a [Dirty] MMR into a [Clean] MMR by computing the digests of any dirty nodes.`